external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaSource.scala

```
@@ -0,0 +1,199 @@
+
+import org.apache.spark.Logging
+import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+import org.apache.spark.sql.execution.streaming.{Batch, Offset, Source, StreamingRelation}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{DataFrame, Dataset, SQLContext}
+
+
+/** An [[Offset]] for the [[KafkaSource]]. */
+private[kafka]
+case class KafkaSourceOffset(offsets: Map[TopicAndPartition, Long]) extends Offset {
+  /**
+    * Returns a negative integer, zero, or a positive integer as this object is less than, equal to,
+    * or greater than the specified object.
+    */
+  override def compareTo(other: Offset): Int = other match {
+    case KafkaSourceOffset(otherOffsets) =>
+      val allTopicAndPartitions = (this.offsets.keySet ++ otherOffsets.keySet).toSeq
+
+      val comparisons = allTopicAndPartitions.map { tp =>
+        (this.offsets.get(tp), otherOffsets.get(tp)) match {
+          case (Some(a), Some(b)) =>
+            if (a < b) {
+              -1
+            } else if (a > b) {
+              1
+            } else {
+              0
+            }
+          case (None, _) =>
+            -1
+          case (_, None) =>
+            1
+          case _ =>
+            throw new IllegalArgumentException(
+              s"Invalid comparison between two sets of Kafka offsets: $this <=> $other")
+        }
+      }
+      val nonZeroSigns = comparisons.filter { _ != 0 }.toSet
+      nonZeroSigns.size match {
+        case 0 => 0 // if both empty or only 0s
+        case 1 => nonZeroSigns.head // if there are only (0s and 1s) or (0s and -1s)
+        case _ => // there are both 1s and -1s
+          throw new IllegalArgumentException(
+            s"Invalid comparison between non-linear histories: $this <=> $other")
+      }
+
+    case _ =>
+      throw new IllegalArgumentException(s"Cannot compare $this <=> $other")
+  }
+
+  override def ==(other: Offset): Boolean = Try(compareTo(other) == 0).getOrElse(false)
```

I would override `equals` since I think that `==` delegates to that.

I also would not hide exceptions, but instead make the matching for `false` explicit.